### PR TITLE
Add option to reading input files from one single yaml file

### PR DESCRIPTION
**Pull request by porya-gohary**
The main changes are:

1. [yaml-cpp](https://github.com/jbeder/yaml-cpp) added as a submodule to the project to parse yaml files
2. Project version updated to 2.2 in [CMakeLists.txt](https://github.com/gnelissen/np-schedulability-analysis/compare/master...porya-gohary:np-schedulability-analysis:add_yaml?expand=1#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a).
3. [CMakeLists.txt](https://github.com/gnelissen/np-schedulability-analysis/compare/master...porya-gohary:np-schedulability-analysis:add_yaml?expand=1#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a) updated to compile yaml-cpp library.
4. In [include/io.hpp](https://github.com/gnelissen/np-schedulability-analysis/compare/master...porya-gohary:np-schedulability-analysis:add_yaml?expand=1#diff-6a4ec05b747b2670559d897d45a42ce4650531740bad9dfdfb1f5983aae53363), two functions `parse_yaml_job_file` and `parse_yaml_dag_file` added to parse job set and their precedence.
5. In [include/io.hpp](https://github.com/gnelissen/np-schedulability-analysis/compare/master...porya-gohary:np-schedulability-analysis:add_yaml?expand=1#diff-6a4ec05b747b2670559d897d45a42ce4650531740bad9dfdfb1f5983aae53363), the old `parse_file` function, renamed to `parse_csv_job_file`.
6. All test cases in `src/tests` were refactored according to changes in 5.
7. Two existing examples `fig1a.csv` (with precedence constraint) and `global-fig.csv` also written in yaml format.

### CODE REVIEW

**Comment by porya-gohary** [comment]
Please notice that in this implementation `yaml-cpp` is used as an integrated library from the source and is not used as an installed library. So after pulling the code, the submodule should also be downloaded using: `git submodule update --init --recursive`. This should be also mentioned in the `README.md`.

**Comment by porya-gohary** [comment]
Now the successors of a job can be written in both compact and expanded formats.
In the compact format, dependencies are described as a sequence. For example:
`Successors: [[Int, Int], [Int, Int]]`
In the expanded format successors are explicitly mentioned as an array. For example:
```
Successors:
           - TaskID: Int
              JobID: Int
           - TaskID: Int
              JobID: Int
```

Note: Both of these syntaxes can be used in one single file together like the example in [examples/fig1a.yaml](https://github.com/gnelissen/np-schedulability-analysis/pull/15/commits/b8c66d235e0d1d7a9561c929c111e38efc34b3f4#diff-e7cf64ad6519ff7d87875298700196537f781a90c5ccbda8ab0cf70fd9f55dc4).

File: include/io.hpp [hunk 1]
```diff
@@ -97,6 +98,52 @@ namespace NP {
...
+		in.seekg(0, std::ios::beg);
+		try {
+			// read the YAML file
+			YAML::Node input_job_set = YAML::Load(in);
+			auto const js = input_job_set["jobset"];
+			// Iterate over each jobset entry
+			for (auto const &j : js) {
+				// Check if a job has a successor
+				if (j["Successors"]) {
+					std::cout << "Job ID: " << j["Job ID"].as<unsigned long>() << std::endl;
```
Comment (gnelissen) [inline include/io.hpp:116]: seems like this line should be removed

File: include/io.hpp [hunk 2]
```diff
@@ -97,6 +98,52 @@ namespace NP {
...
+					auto from = JobID(j["Job ID"].as<unsigned long>(), j["Task ID"].as<unsigned long>());
+					// Iterate over each successor
+					for (const auto &succ: j["Successors"]) {
+						// first, we need to check to see if it is written
+						// in the compact form [TaskID, JobID]
+						// or the expanded form
+						// - Task ID: Int
+						// 	 Job ID: Int
+						if (succ.IsSequence()) {
+							std::cout << "Task ID: " << succ[0].as<unsigned long>() << std::endl;
```
Comment (gnelissen) [inline include/io.hpp:126]: same her. The cout should be removed

File: include/io.hpp
Comment (gnelissen) [inline include/io.hpp:1]: a couple of cout must be cleaned

**Review by gnelissen — CHANGES_REQUESTED** [review]
clean io.hpp from unneeded printing

**Pull request merged**